File: service-notification/src/main/java/com/acme/services/notification/java/NotificationServiceImpl.java
package com.acme.services.notification.java;


import com.acme.contracts.services.invoice.Invoice;
import com.acme.contracts.services.notification.BookingRequest;
import com.acme.contracts.services.notification.NotificationService;
import com.acme.common.AbstractService;
import com.acme.contracts.workflows.invoicing.User;

public class NotificationServiceImpl extends AbstractService implements NotificationService {
    @Override
    public void notifyHostOfRequest(BookingRequest request, int attempt) {
        log("Request " + request.getId() + ": Notifying Host " + request.getHost() + " of a new request from " + request.getTraveler() + " (attempt " + attempt + ")");
    }

    @Override
    public void notifyHostOfRequestExpiration(BookingRequest request) {
        log("Request " + request.getId() + ": Notifying Host " + request.getHost() + " of the expiration of the request");
    }

    @Override
    public void notifyTravelerOfRequestExpiration(BookingRequest request) {
        log("Request " + request.getId() + ": Notifying Traveler " + request.getTraveler() + " of the expiration her request");
    }

    @Override
    public void notifyHostOfPaymentSuccess(BookingRequest request) {
        log("Request " + request.getId() + ": Notifying Host " + request.getHost() + " of a successful payment");
    }

    @Override
    public void notifyTravelerOfBookingSuccess(BookingRequest request) {
        log("Request " + request.getId() + ": Notifying Traveler " + request.getTraveler() + " of the success of her booking");
    }

    @Override
    public void notifyTravelerOfBookingDenial(BookingRequest request) {
        log("Request " + request.getId() + ": Notifying Traveler " + request.getTraveler() + " of the denial of her booking");
    }

    @Override
    public void sendInvoice(User user, Invoice invoice) {
        log("Send Invoice " + invoice.getId() + " to user " + user.getId());
    }
}

File: service-notification/src/main/kotlin/com/acme/services/notification/kotlin/NotificationServiceImpl.kt
package com.acme.services.notification.kotlin

import com.acme.contracts.services.invoice.Invoice
import com.acme.contracts.services.notification.BookingRequest
import com.acme.contracts.services.notification.NotificationService
import com.acme.common.AbstractService.log
import com.acme.contracts.workflows.invoicing.User


class NotificationServiceImpl: NotificationService {
    override fun notifyHostOfRequest(request: BookingRequest, attempt: Int) {
        log("Request ${request.id}: Notifying Host ${request.host} of a new request from ${request.traveler} (attempt $attempt)")
    }

    override fun notifyHostOfRequestExpiration(request: BookingRequest) {
        log("Request ${request.id}: Notifying Host ${request.host} of the expiration the request")
    }

    override fun notifyTravelerOfRequestExpiration(request: BookingRequest) {
        log("Request ${request.id}: Notifying Traveler ${request.traveler} of the expiration her request")
    }

    override fun notifyHostOfPaymentSuccess(request: BookingRequest) {
            log("Request ${request.id}: Notifying Host ${request.host} of a successful payment")
    }

    override fun notifyTravelerOfBookingSuccess(request: BookingRequest) {
        log("Request ${request.id}: Notifying Traveler ${request.traveler} of the success of her booking")
    }

    override fun notifyTravelerOfBookingDenial(request: BookingRequest) {
        log("Request ${request.id}: Notifying Traveler ${request.traveler} of the denial of her booking")
    }

    override fun sendInvoice(user: User, invoice: Invoice) {
        log("Send Invoice ${invoice.id} to user ${user.id}")
    }
}
